src/orchestrator.py

```diff
@@ -18,58 +21,92 @@ def add_robot(self, robot_name, initial_pose, end_pose):
+        self.robots[robot_name] = Robot(
+            robot_name=robot_name,
+            obstacles=self.shelves,
+            charge_locations=self.charge_locations,
+            orchestrator=self,
+            max_x=self.size[0],
+            max_y=self.size[1],
+            initial_pose=initial_pose,
+            end_pose=end_pose
+)
+        self.waiting_robots.add(robot_name)
+        # self.robots[robot_name].plan_path()
+        # first, second = self.robots[robot_name].get_next_two_points()
+        # self.lock_cells(self.robots[robot_name], first, second)
         return self.robots[robot_name]
 
+    def make_request(self, end_pose):
+        self.request_queue.append(end_pose)
+
+    def on_deadlock(self, pose):
+        self.deadlock_count += 1
+        for observer in self.deadlock_observers:
+            observer.__call__(pose)
+
     def move_all(self):
         """
         Execute a move for all of the robots under the orchestrator
         control. If a collision is detected, replan the path.
         """
-        robots_to_move = [self.robots[r] for r in self.robots if not self.robots[r].is_done()]
-        for robot in robots_to_move:
+        robots_to_move = [(r, self.robots[r]) for r in self.robots if not self.robots[r].is_done()]
+        for id, robot in robots_to_move:
             # unlock reserved pts
             for pt in robot.locked_cells:
                 if pt in self.locked:
                     self.locked.remove(pt)
-            robot.locked_cells.clear()
-
-            # move the robot to the next pose
-            robot.move_robot()
-            # if arrived at goal, we're done!
-            if robot.is_done():
-                continue
+            robot.locked_cells.clear()  
+
+            # check if the robot is waiting
+            # for a path. If so, try to plan
+            # if we still can't find a path, skip this iteration
+            if id in self.waiting_robots:
+                success = robot.plan_path()
+                if success:
+                    self.waiting_robots.remove(id)
+                else:
+                    self.on_deadlock(robot.current_pose)
+                    continue
 
             # identify the next two pts we need to
             # lock for this robot
             first, second = robot.get_next_two_points()
-
             # if either is already reserved for another robot
             # we need to replan the path
-            if self.is_pt_locked(first) or self.is_pt_locked(second):
-                self.deadlock_count += 1
-                for observer in self.deadlock_observers:
-                    observer.__call__(robot.current_pose)
-                robot.plan_path()
-                first, second = robot.get_next_two_points()
-
+            if self.is_pt_locked(first) or self.is_pt_locked(second):                
+                self.on_deadlock(robot.current_pose)
+                self.waiting_robots.add(id)
```

Comment:
so, `waiting_robots` is not a list of robots without goals, but robots that need a path planned?

Reply:
If this is true, can we put a small comment on L14 for this?

Reply:
Yeah, you got it. I can add a comment.